### PR TITLE
Updated Decaf Guide

### PR DESCRIPTION
**Pull request by confused-Techie**
Since there has been more talk about how to decaf for Pulsar, DeeDeeG and I thought it'd be best if there was a fully up to date guide on how to do so. So that no longer is the steps here just whoever answers first, but is a proper guide that we can all try to stick to.

This guide is created based on the steps that have evolved over time to do decaf work, and is based on the most successful and easy steps I've yet to find to reliably do any decaf work.

Please feel free to suggest any edits as needed.

I also thought it'd be a good idea, at least until we come up with something better, but we can start to use this new `guide` folder as the location we upload any guides that span across the whole org. For instance the publishing guide/checklist would be a fantastic thing to also put here. Which if we like that idea I might just get on that tonight.

### CODE REVIEW

File: guides/how-to-decaf.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+This section covers the general overview and workflow you should adopt to successfully decaf some code.
+
+But first to do so, there is one big rule:
+
+  * No matter what, any PR that decafs **both** the source files of a repo/package **and** the specs cannot be accepted. Since decaffing code can be tricky, we need to ensure we catch any errors that are being produced. If both the specs are decaffed at the same time the source files are, it is impossible to tell if the process happened successfully, and which aspect is producing any test failures or errors we see later on. Generally, if a repo/package needs both the specs and source files decaffed, they should each occur as their own unique PRs.
+
+Secondly, here is a general recommendation when choosing what to decaf:
+
+  * When you decide to decaf a repo/package, it's a good idea to decaf the entirety of what you are working on. Decaffeinating only a single spec file or single source file doesn't do us much good, and in many cases may be totally invalid on it's own, so it's highly recommended to, within a single PR, decaf the entire source files, or entire spec files.
```
Comment (Daeraxa) [inline guides/how-to-decaf.md:32]:
```suggestion
  * When you decide to decaf a repo/package, it's a good idea to decaf the entirety of what you are working on. Decaffeinating only a single spec file or single source file doesn't do us much good, and in many cases may be totally invalid on its own, so it's highly recommended to, within a single PR, decaf the entirety of the source or spec files.
```

File: guides/how-to-decaf.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+With those rules/recommendations out of the way, following below is the general guide that should be followed:
```
Comment (Daeraxa) [inline guides/how-to-decaf.md:34]:
```suggestion
With those rules/recommendations out of the way, below is the general guide that should be followed:
```

File: guides/how-to-decaf.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+1. As you might expect, decide what you are going to decaffeinate. To help make everyone's lives easier, you should ensure the file you are decaffing doesn't have any open PRs against it, or be at risk of receiving major changes anytime soon. If you aren't sure of this, feel free to ask the Pulsar maintainers.
```
Comment (Daeraxa) [inline guides/how-to-decaf.md:36]:
```suggestion
1. As you might expect, decide what you are going to decaffeinate. To help make everyone's lives easier, you should ensure the file you are decaffing doesn't have any open PRs against it, or that it is at risk of receiving major changes anytime soon. If you aren't sure of this, feel free to ask the Pulsar maintainers.
```

File: guides/how-to-decaf.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+2. Once you've decided what repo/package you are going to decaf, then decide if you want to decaf the specs or source files.
+
+3. Now with the choices out of the way, you can get started. You'll want to make sure to create the Machine Decaf branch (more on that later) but this would be a branch that exists on the remote repo that contains only machine decaffed files. (If you don't have permissions to make this, feel free to ask the Pulsar maintainers). While we could be done there, the machine decaffed code is valid, but very unreadable and difficult to work with, so it requires manual intervention to improve.
```
Comment (Daeraxa) [inline guides/how-to-decaf.md:40]: This seems confusing to me, are we saying that the machine decaf branch exists on the original repo before one can start? This seems somewhat difficult for people to get involved with.

File: guides/how-to-decaf.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+## Machine Decaf
+
+Like mentioned earlier, using automated tools to decaf CoffeeScript produces valid, yet unreadable and unmaintainable code.
+
+To produce this new code in a way that we know is functional, you'll want to keep lookout for a few things:
+
+First note, that in general using CoffeeScript 2.x (the latest version as of this writing) is totally valid. But if you find any tests are failing with only machine decaf code, it may be because there are very few edge cases where CoffeeScript code decafs differently from 2.x tooling and 1.x tooling. Which because nearly all the CoffeeScript for Atom was written back in the CoffeeScript 1.x days, we need to account for this. You can always check if a particular repo/package is using 1.x CoffeeScript tooling by checking the version of `coffee-script` or `coffeescript` within the relevant `package.json` either `dependencies` or `devDependencies`. If those are within the 1.x range, you should use older tooling to decaf. While keeping in mind that the 1.x CoffeeScript decaf tooling will usually strip any comments from the code, and will have to be added back in.
```
Comment (Daeraxa) [inline guides/how-to-decaf.md:54]:
```suggestion
First note, that in general using CoffeeScript 2.x (the latest version as of this writing) is totally valid. But if you find any tests are failing with only machine decaf code, it may be because there are very few edge cases where CoffeeScript code decafs differently from 2.x tooling and 1.x tooling. Which, because nearly all the CoffeeScript for Atom was written back in the CoffeeScript 1.x days, we need to account for. You can always check if a particular repo/package is using 1.x CoffeeScript tooling by checking the version of `coffee-script` or `coffeescript` within the relevant `package.json` either `dependencies` or `devDependencies`. If those are within the 1.x range, you should use older tooling to decaf. While keeping in mind that the 1.x CoffeeScript decaf tooling will usually strip any comments from the code, and will have to be added back in.
```

**Review by Daeraxa — COMMENTED** [review]
Just a bunch of small corrections and readability updates

File: guides/how-to-decaf.md (hunk 4 above)
Comment (confused-Techie) [reply guides/how-to-decaf.md:40]: While I know that's a hard barrier to entry it does. 
The remote repository needs to have a machine decaf branch first, so that the contributor doing the manual decaf work is able to create a branch from those changes (so they have the machine decaf), and more importantly, it's the branch they target to merge their manual decaf. 

Since, there's no way to target a branch that doesn't exist on remote in a PR that's why it needs to already exist. Since if we don't have a machine decaf and manual decaf diff, these are impossible to review.

That's why I mentioned that if the contributor doesn't have permissions that they can contact a Pulsar maintainer to have that branch made, as this is how I've been doing all my decaf work, and how it's being done for PPM. Before the latest PPM decaf PRs were made I had created a machine decaf branch for the contributor to target.

Comment (Daeraxa) [reply guides/how-to-decaf.md:40]: Ok I get you, I was thinking that an alternative is to submit a PR to create a machine decaf branch as a separate PR from the manual work but thinking back I realise that isn't possible without having access first as you can only compare branches in a PR, not target the creation of a new one.

Comment (confused-Techie) [reply guides/how-to-decaf.md:40]: Yeah, I wish there was a better way to do this.

We could possibly instruct to create the machine decaf as a single commit, then only commits afterwards can add manual decaf steps. And we review a PR based on the diff between commits, but that sounds like it's much more likely to be done improperly, but even so not sure that's worth the added ability of anyone being able to work on it, without having to wait on us.

Comment (Daeraxa) [reply guides/how-to-decaf.md:40]: I agree, I think what you suggest is probably best.

**Review by Daeraxa — APPROVED** [review]
So this approval from me is simply around the documentation side - not the actual content. i.e. I'm assuming all the actual procedures and advice given here are correct.

Looks great assuming it is all good on that front as well.

**Comment by confused-Techie** [comment]
@Daeraxa Thanks for the approval, I'll ad this one in
